backend/crawlers/ddc.js

```diff
@@ -0,0 +1,167 @@
+    return;
+  }
+
+  let metrics;
+  try {
+    const response = await axios.get(`${url}/metrics/stats`);
+    metrics = response.data;
+  } catch (error) {
+    logger.error(loggerOptions, `Can't collect storage node metrics ${node.node_id} on ${url} (${error.toString()})`);
+    return;
+  }
+  const dataStoredMetricId = await insertMetricAndReturnId(dbClient, 'dataStoredBytes', node.node_id);
+  const piecesStoredMetricId = await insertMetricAndReturnId(dbClient, 'piecesStored', node.node_id);
+  await insertMetricValue(dbClient, dataStoredMetricId, metrics.dataStoredBytes);
+  await insertMetricValue(dbClient, piecesStoredMetricId, metrics.piecesStored);
+}
+
+async function collectStorageMetrics(dbClient, nodes) {
+  logger.info(loggerOptions, 'Collecting storage metrics...');
+  return Promise.all(nodes.map((n) => collectStorageNodeMetrics(dbClient, n)));
+}
+
+async function collectGatewayNodeMetrics(dbClient, node) {
+  const { url } = JSON.parse(node.params);
+  if (!url) {
+    logger.info(`Can't collect gateway node ${node.node_id} metrics (url undefined)`);
+    return;
+  }
+
+  let metrics;
+  try {
+    const response = await axios.get(`${url}/metrics/stats`);
+    metrics = response.data;
+  } catch (error) {
+    logger.error(loggerOptions, `Can't collect gateway node metrics ${node.node_id} (${JSON.stringify(error)})`);
+    return;
+  }
+
+  const piecesViewedMetricId = await insertMetricAndReturnId(dbClient, 'piecesViewed', node.node_id);
+  const avgResponseTimeMetricId = await insertMetricAndReturnId(dbClient, 'avgResponseTimeSec', node.node_id);
+  const avgDownloadSpeedMetricId = await insertMetricAndReturnId(dbClient, 'avgDownloadSpeedBytesPerSec', node.node_id);
+  const avgUploadSpeedMetricId = await insertMetricAndReturnId(dbClient, 'avgUploadSpeedBytesPerSec', node.node_id);
+  await insertMetricValue(dbClient, piecesViewedMetricId, metrics.piecesViewed);
+  await insertMetricValue(dbClient, avgResponseTimeMetricId, metrics.avgResponseTimeSec);
+  await insertMetricValue(dbClient, avgDownloadSpeedMetricId, metrics.avgDownloadSpeedBytesPerSec);
+  await insertMetricValue(dbClient, avgUploadSpeedMetricId, metrics.avgUploadSpeedBytesPerSec);
+}
+
+async function collectGatewayMetrics(dbClient, nodes) {
+  logger.info(loggerOptions, 'Collecting gateway metrics...');
+  return Promise.all(nodes.map((n) => collectGatewayNodeMetrics(dbClient, n)));
+}
+
+const crawler = async () => {
+  logger.info(loggerOptions, 'Starting DDC crawler...');
+  const dbClient = await getClient(loggerOptions);
+
+  const { api, chainName } = await connect(config.rpc);
+
+  const contract = getContract(config.contractName, chainName, api);
+
+  const { storageNodes, gatewayNodes } = await collectContractMetrics(dbClient, contract);
+  await collectStorageMetrics(dbClient, storageNodes);
+  await collectGatewayMetrics(dbClient, gatewayNodes);
+
+  logger.info(loggerOptions, 'DDC crawler completed');
```

Comment:
Should crawler up periodically? If yes, where is this logic?

Reply:
added